src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdmBuilder.java

```diff
@@ -0,0 +1,448 @@
+    }
+
+    private static String getVastImpClickErrorTrackingUrls(List<String> urls, MonitorEventType eventType) {
+        return urls.stream().map(url -> makeUrl(eventType, url)).collect(Collectors.joining());
+    }
+
+    private static String makeUrl(MonitorEventType eventType, String url) {
+        return switch (eventType) {
+            case CLICK -> "<ClickTracking><![CDATA[" + url + "]]></ClickTracking>";
+            case IMP -> "<Impression><![CDATA[" + url + "]]></Impression>";
+            case VAST_ERROR -> "<Error><![CDATA[" + url + "&et=[ERRORCODE]]]></Error>";
+            default -> "";
+        };
+    }
+
+    private static String getVastEventTrackingUrls(List<String> urls, MonitorEventType eventType) {
+        return urls.stream().map(eventUrl -> eventType == MonitorEventType.USER_CLOSE
+                        ? "<Tracking event=\"skip\"><![CDATA[" + eventUrl + "]]></Tracking>"
+                        + "<Tracking event=\"closeLinear\"><![CDATA[" + eventUrl + "]]></Tracking>"
+                        : "<Tracking event=\"" + eventType.getEvent() + "\"><![CDATA[" + eventUrl + "]]></Tracking>")
+                .collect(Collectors.joining());
+    }
+
+    private String buildRewardedVideoPart(Content content, Integer adWidth, Integer adHeight, String clickUrl) {
+        final MetaData metaData = content.getMetaData();
+
+        final RewardedVideoPartAdm.RewardedVideoPartAdmBuilder admBuilder = RewardedVideoPartAdm.builder()
+                .id(content.getContentId())
+                .adId(content.getContentId())
+                .clickUrl(clickUrl);
+
+        final List<Icon> iconList = metaData.getIconList();
+        if (CollectionUtils.isNotEmpty(iconList) && StringUtils.isNotBlank(iconList.get(0).getUrl())) {
+            final Icon icon = iconList.get(0);
+            final boolean isFormatDefined = HuaweiUtils.isFormatDefined(icon.getWidth(), icon.getHeight());
+            return admBuilder
+                    .staticImageUrl(icon.getUrl())
+                    .staticImageHeight(isFormatDefined ? icon.getHeight() : adHeight)
+                    .staticImageWidth(isFormatDefined ? icon.getWidth() : adWidth)
+                    .build().toString();
+        }
+
+        final List<ImageInfo> imageInfoList = metaData.getImageInfoList();
+        if (CollectionUtils.isNotEmpty(imageInfoList) && StringUtils.isNotBlank(imageInfoList.get(0).getUrl())) {
+            final ImageInfo imageInfo = imageInfoList.get(0);
+            final boolean isFormatDefined = HuaweiUtils.isFormatDefined(imageInfo.getWidth(), imageInfo.getHeight());
+            return admBuilder
+                    .staticImageUrl(imageInfo.getUrl())
+                    .staticImageHeight(isFormatDefined ? imageInfo.getHeight() : adHeight)
+                    .staticImageWidth(isFormatDefined ? imageInfo.getWidth() : adWidth)
+                    .build().toString();
+        }
```

Comment:
This two parts similar. Please, create method for this logic and use it for this two cases.

Reply:
it will be cumbersome, because ImageInfo and Icon are not under the same hierarchy